presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayOfRowUnnestBlockBuilder.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.unnest;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.type.Type;
+
+import static com.facebook.presto.operator.unnest.NullElementFinder.INVALID_NULL_POSITION;
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+class ArrayOfRowUnnestBlockBuilder
```

Comment:
It seems to me that you don't need this class at all. Just add "optional" nullCheckBlock to UnnestBlockBuilder

```
    public Block buildOutputBlockWithNulls(int[] maxEntries, int offset, int positionCount, int batchSize)
    {
        int nullIndex = nullFinder.getNullElementIndex();
        if (nullIndex == INVALID_NULL_POSITION) {
            source = insertNull(source, type);
            nullIndex = source.getPositionCount() - 1;
            nullFinder.setNullElementIndex(nullIndex);
        }

        int position = 0;
        int[] ids = new int[batchSize];
        for (int i = 0; i < positionCount; i++) {
            int length = lengths[offset + i];
            for (int j = 0; j < length; j++) {
                if (nullCheckBlock != null && nullCheckBlock.isNull(nullCheckBlockPosition++)) {
                    ids[position++] = nullIndex;
                }
                else {
                    ids[position++] = sourcePosition++;
                }
            }

            int maxLength = maxEntries[offset + i];
            for (int j = length; j < maxLength; j++) {
                ids[position++] = nullIndex;
            }
        }

        return new DictionaryBlock(batchSize, source, ids);
    }
```

Reply:
The reason behind it was trying to avoid if checks in the loop for GenericUnnestBlockBuilder so the loop could be optimized. Mixing the two loops would regress the GenericUnnestBlockBuilder by over 20%.

Reply:
@yingsu00 Ok. Let's not mix the loops. Just move the if check outside of the loop. I still don't think we need a separate class and would like to avoid getSourcePosition/setSourcePosition pattern.